communication/src/v2/coap_payload.cpp

```diff
@@ -0,0 +1,215 @@
+    size_t curPos = CHECK_FS(lfs_file_tell(fs, file));
+    // Prevent LittleFS from flushing the file if the offset hasn't changed
+    if (pos != curPos) {
+        CHECK_FS(lfs_file_seek(fs, file, pos, LFS_SEEK_SET));
+    }
+    return 0;
+}
+
+inline int closeFile(lfs_t* fs, lfs_file_t* file) {
+    CHECK_FS(lfs_file_close(fs, file));
+    return 0;
+}
+
+inline int removeFile(lfs_t* fs, const char* path) {
+    CHECK_FS(lfs_remove(fs, path));
+    return 0;
+}
+
+} // namespace
+
+CoapPayload::CoapPayload(size_t maxHeapSize) :
+        dataSize_(0),
+        maxHeapSize_(std::min<size_t>(maxHeapSize, COAP_MAX_PAYLOAD_SIZE)),
+        fileNum_(0) {
+}
+
+CoapPayload::~CoapPayload() {
+    if (file_) {
+        FsLock fs;
+        removeTempFile(fs.instance());
+    }
+}
+
+int CoapPayload::read(char* data, size_t size, size_t pos) {
+    if (pos >= dataSize_) {
+        return SYSTEM_ERROR_END_OF_STREAM;
+    }
+    char* d = data;
+    size_t bytesToRead = std::min(size, dataSize_ - pos);
+    if (pos < maxHeapSize_) {
+        size_t bytesInRam = std::min(dataSize_, maxHeapSize_) - pos;
+        size_t n = std::min(bytesToRead, bytesInRam);
+        std::memcpy(d, buf_.data() + pos, n);
+        bytesToRead -= n;
+        pos += n;
+        d += n;
+    }
+    if (bytesToRead > 0) {
+        FsLock fs;
+        assert(file_);
+        CHECK(seekInFile(fs.instance(), file_.get(), pos - maxHeapSize_));
+        size_t n = CHECK_FS(lfs_file_read(fs.instance(), file_.get(), d, bytesToRead));
+        if (n < bytesToRead) {
+            // The payload size was set using setSize() or the file was modifed by somebody else
+            std::memset(d + n, 0, bytesToRead - n);
+        }
+        d += bytesToRead;
+    }
+    return d - data;
+}
+
+int CoapPayload::write(const char* data, size_t size, size_t pos) {
+    if (pos + size > COAP_MAX_PAYLOAD_SIZE) {
+        return SYSTEM_ERROR_COAP_TOO_LARGE_PAYLOAD;
+    }
+    size_t p = pos;
+    size_t bytesToWrite = size;
+    if (p < maxHeapSize_) {
+        size_t n = std::min(bytesToWrite, maxHeapSize_ - p);
+        auto newSize = buf_.size() + n;
+        if (newSize > buf_.capacity()) {
+            // TODO: Use a pool of fixed-size chainable buffers
+            auto newCapacity = std::min(std::max(std::max(buf_.capacity() * 3 / 2, newSize), INITIAL_BUFFER_CAPACITY), maxHeapSize_);
+            if (!buf_.reserve(newCapacity)) {
+                return SYSTEM_ERROR_NO_MEMORY;
+            }
+        }
+        buf_.resize(newSize);
+        std::memcpy(buf_.data() + p, data, n);
+        bytesToWrite -= n;
+        p += n;
+        data += n;
+    }
+    if (bytesToWrite > 0) {
+        FsLock fs;
+        if (!file_) {
+            CHECK(createTempFile(fs.instance()));
+        }
+        assert(file_);
+        CHECK(seekInFile(fs.instance(), file_.get(), p - maxHeapSize_));
+        size_t n = CHECK_FS(lfs_file_write(fs.instance(), file_.get(), data, bytesToWrite));
```

Comment:
todo: make sure that lfs_file_write can handle writing large buffers (>1000 bytes) in one shot or split the writes in here to the limit of lfs_file_write

Reply:
@avtolstoy Is this still something on your team's radar or was it perhaps fixed already? It's that problem with writing to external flash when the source buffer is in internal flash, possibly at an unaligned address.

Reply:
I have a fix for this, should get a PR out today.

Reply:
Thanks!